src/ansi-c/library/cprover_contracts.c

```diff
@@ -0,0 +1,4116 @@
+
+typedef __CPROVER_contracts_car_set_t *__CPROVER_contracts_car_set_ptr_t;
+
+typedef struct __CPROVER_contracts_obj_set_t
+{
+  // maximum number of elements that can be stored in the set
+  __CPROVER_size_t max_elems;
+  // 1 + greatest used index in elems,
+  // next usable index in elems if less than max_elems
+  __CPROVER_size_t watermark;
+  // number of elements currently in the elems array
+  __CPROVER_size_t nof_elems;
+  // true iff nof_elems is 0
+  __CPROVER_bool is_empty;
+  // true iff elems is indexed by the object id of the pointers
+  __CPROVER_bool indexed_by_object_id;
+  // array of void *pointers, indexed by their object ID
+  // or some other order
+  void **elems;
+} __CPROVER_contracts_obj_set_t;
+
+typedef __CPROVER_contracts_obj_set_t *__CPROVER_contracts_obj_set_ptr_t;
+
+typedef struct __CPROVER_contracts_write_set_t
+{
+  // set of car derived from the contract (by rank)
+  __CPROVER_contracts_car_set_t contract_assigns;
+  // set of freeable pointers derived from the contract (by object id)
+  __CPROVER_contracts_obj_set_t contract_frees;
+  // set of freeable pointers derived from the contract (by rank)
+  __CPROVER_contracts_obj_set_t contract_frees_replacement;
+  // set of objects allocated by the function under analysis (by object id)
+  __CPROVER_contracts_obj_set_t allocated;
+  // set of objects deallocated by the function under analysis (by object id)
+  __CPROVER_contracts_obj_set_t deallocated;
+  // object set supporting the is_freshr predicate checks (by object id)
+  __CPROVER_contracts_obj_set_t is_freshr_seen;
+  // object set recording the is_freshr allocations in post conditions
+  // (replacement mode only)
+  __CPROVER_contracts_obj_set_ptr_t linked_allocated;
+  // true iff this write set is used for contract replacement
+  __CPROVER_bool replacement;
+  // true iff this write set checks requires clauses in an assumption ctx
+  __CPROVER_bool assume_requires_ctx;
+  // true iff this write set checks requires clauses in an assertion ctx
+  __CPROVER_bool assert_requires_ctx;
+  // true iff this write set checks ensures clauses in an assumption ctx
+  __CPROVER_bool assume_ensures_ctx;
+  // true iff this write set checks ensures clauses in an assertion ctx
+  __CPROVER_bool assert_ensures_ctx;
+} __CPROVER_contracts_write_set_t;
+
+typedef __CPROVER_contracts_write_set_t *__CPROVER_contracts_write_set_ptr_t;
+#endif
+
+int __builtin_clzll(unsigned long long);
+
+inline void __CPROVER_contracts_obj_set_create_indexed_by_object_id(
+  __CPROVER_contracts_obj_set_ptr_t set)
+{
+  __CPROVER_assert(
+    __CPROVER_rw_ok(set, sizeof(__CPROVER_contracts_obj_set_t)),
+    "set writable");
+  // compute the maximum number of objects that can exist in the
+  // symex state from the number of object_bits/offset_bits
+  __CPROVER_size_t object_bits = __builtin_clzll(__CPROVER_max_malloc_size);
+  __CPROVER_size_t nof_objects = 1UL << object_bits;
+  set->max_elems = nof_objects;
+  set->watermark = 0;
```

Comment:
I think I prefer "true" and "false" to "1" and "0" to emphasize we're talking about booleans.